Moving active span interaction To Trace Context Utilities

toumorokoshi opened the pull request:
Fixes #924

## Motivation

The Trace API currently

## Changes

The previous spec stated that the active span get / set behavior MUST
behave the same across Tracers retrieved from the same TracerProvider.
As such, the active span manipulation methods are better provided by the
TracerProvider directly.

As it is still cumbersome to retrieve the current TracerProvider simply
to get the current span, add in the concept of a "Trace Package", that can
provide utilty methods to delegate to the configured default TracerProvider.
This enables a fluid interface to retrieve the current span:

    from opentelemetry import trace

    current_span = trace.get_current_span()

To help maintain backwards compatibility, the Tracer may still expose
active span methods.

yurishkuro commented on specification/trace/api.md:
```diff
@@ -74,6 +75,18 @@ A duration is the elapsed time between two events.
+## Trace package
+
+The `Trace Package` is the package containing the TracerProvider and Tracer
+API classes. In some languages, this may also be referred to as a module.
```
what would this mean for Java, where a package by itself cannot have functions?

Oberon00 replied:
Actually Java already implements it this way, it has a class with static member functions, see https://github.com/open-telemetry/opentelemetry-java/blob/v0.8.0/api/src/main/java/io/opentelemetry/trace/TracingContextUtils.java#L31

toumorokoshi replied:
That's what I was thinking too. I'll add that to the PR.

toumorokoshi replied:
I've added a section, please re-open if that doesn't address this comment sufficiently.

yurishkuro reviewed (changes requested):
I don't understand the motivation for this. Span interactions, including active span, are best done via Tracer, it minimizes the number of objects that instrumentation has to deal with. How that's done by the tracer is an implementation detail, the only requirement is for active span accessors to act consistently.

And it does not fix #455, where the objective was to achieve the most normalized (albeit more verbose) API, which are maximally decoupled (i.e. Tracer does not need to know anything about Context - see opentracing.Tracer, for example).

Oberon00 commented on specification/trace/api.md:
```diff
@@ -74,6 +75,18 @@ A duration is the elapsed time between two events.
+- Get the currently active span
+- Set the currently active span
+
+Both of these operations MUST delegate to the configured `TracerProvider`.
```
Why does the TracerProvider need to be involved? I don't think we should encourage having a per-TracerProvider active span. There should be only one per context normally.

toumorokoshi replied:
I agree, but the [original wording](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#tracer-operations) implies that this can be the case:

```
In particular, this implies that the active span MUST not depend on the Tracer that it is queried from/was
set to, as long as the tracers were obtained from the same TracerProvider.
```

That choice may have more ramifications (e.g. 3rd-party tracer providers would not be able to directly modify *how* they work with the context management system), but if you feel there's merit, I can file a separate issue / PR, or just remove that wording in this one.

toumorokoshi replied:
Based on this conversation, I'm going to move these operations up to the package level, which implicitly results in vendor TracerProviders not being able to modify the behavior of how these spans are set or retrieved.

toumorokoshi commented:
@yurishkuro regarding #455: yes, it's a good point that the original point of the issue isn't solved. I've added my rationale and moved the references to a new issue #924 that describes the desire in more detail.

Regarding who should perform span interactions: creating a Span currently requires a Tracer to ensure named tracers, so that certainly can't be moved. 

But there is a common use case of retrieving the active span, for the purposes of adding or modifying it in some fashion. The API as written implies that the active span should be the same regardless of which Tracer retrieves it. As such, retrieving the span from a tracer has a few cons, from my perspective:

- allocating an otherwise unused Tracer if the use case is just modifying the active span
- unclear to a user how the tracer name impacts that calculation for the active span (unless you have a strong understanding of how tracing works). 

APIs are subjective, but [adding get_current_span to the module](https://github.com/open-telemetry/opentelemetry-python/pull/552) in Python had a quorum that this was a change that improved the user experience significantly.

carlosalberto commented:
@toumorokoshi Thanks for the PR. As @Oberon00 mentioned,  this is better separated from `Tracer`, but it doesn't need to be in `TracerProvider` neither. As any `Tracer` / `TracerProvider` can consume whatever is in the current `Context`, I'd put it in its separate class. 

Your example would stay the same (i.e. `trace.current_span()`) as the active `Span` would depend on the top level package for tracing (`opentelemetry.trace`) which is fine ;)

> To help maintain backwards compatibility, the Tracer may still expose
active span methods.

In OT we had these as shortcuts, but I heard some users got confused by it. Something to ponder.

toumorokoshi commented:
> @toumorokoshi Thanks for the PR. As @Oberon00 mentioned, this is better separated from `Tracer`, but it doesn't need to be in `TracerProvider` neither. As any `Tracer` / `TracerProvider` can consume whatever is in the current `Context`, I'd put it in its separate class.

Ok! That sounds like another opinion supporting https://github.com/open-telemetry/opentelemetry-specification/pull/923#discussion_r484189776, so I'll get that PR up as well.

> In OT we had these as shortcuts, but I heard some users got confused by it. Something to ponder.

@carlosalberto would you mind elaborating? I think we still need to leave those methods for backwards compatibility, but would be interested in more data to inform the final design.

toumorokoshi commented:
@Oberon00 @carlosalberto addressed your concerns, PTAL.

Oberon00 commented on specification/trace/api.md:
```diff
@@ -74,6 +75,21 @@ A duration is the elapsed time between two events.
 * The minimal precision is milliseconds.
 * The maximal precision is nanoseconds.
 
+## Trace Package
```
I would prefer to use "module" as the more generic term.

toumorokoshi replied:
At Carlos' suggestion, I've explicitly called out a Class (Tracing Context Utilities) with a caveat that enables it to be a static module function instead. Please take a look.

carlosalberto commented on spec-compliance-matrix.md:
```diff
@@ -11,6 +11,9 @@ status of the feature is not known.
 
 |Feature                                       |Go |Java|JS |Python|Ruby|Erlang|PHP|Rust|C++|.Net|
 |----------------------------------------------|---|----|---|------|----|------|---|----|---|----|
+|[Trace Package](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#trace-package)|
```
I'd rename this to `Trace Context Component` or similar, as it will be exposed as a class, rather than a module, in languages such as .Net or Java.

Also I'd put it below `TracerProvider`, which is supposed to be the root functionality component in the tracing section ;)

toumorokoshi replied:
done!

carlosalberto commented:
> would you mind elaborating?

So in OT we had the context functionality in a `ScopeManager` class, which would expose `active()` and `activateSpan()` methods; `Tracer` in turn would have an `activeSpan()` method that was simply a shortcut to the respective `ScopeManager` method.

But to simplify things, lets consider removing such `Tracer` calls in a follow up, to not delay this PR ;)

Oberon00 commented on specification/trace/api.md:
```diff
@@ -97,7 +98,7 @@ number of `TracerProvider` instances.
 The `TracerProvider` MUST provide functions to:
 
-- Get a `Tracer`
+#### Get a `Tracer`
```
Update the TOC? Maybe switch the toc to markdown-toc.

toumorokoshi replied:
I can revert for now. It's more correct IMO, but unrelated to the change.

toumorokoshi replied:
reverted!

carlosalberto commented:
Thanks, looks great! @open-telemetry/specs-approvers Please review.

Oberon00 commented:
@yurishkuro
> Span interactions, including active span, are best done via Tracer, it minimizes the number of objects that instrumentation has to deal with. How that's done by the tracer is an implementation detail,

That might be a good design too, but in OpenTelemetry, the Tracer MUST use the Context system, that is even specified in the API spec.

> The Tracer MUST delegate to the [Context](https://github.com/open-telemetry/opentelemetry-specification/blob/a0828f6223f7b25aaba567179575c48e942bcead/specification/context/context.md) to perform these tasks, i.e. the above methods MUST do the same as a single equivalent method of the Context management system.

-- https://github.com/open-telemetry/opentelemetry-specification/blob/a0828f6223f7b25aaba567179575c48e942bcead/specification/trace/api.md#tracer-operations

bogdandrutu reviewed (approved):
Thanks for taking care of this.

carlosalberto commented:
Ping @yurishkuro - all feedback has addressed, I think, but would love to have our final review (take?) on this before we actually merge it ;)

bogdandrutu commented:
@toumorokoshi please rebase

arminru commented:
@bogdandrutu I updated the branch (just a conflict in the changelog) but we're still waiting for @yurishkuro to approve (or dismiss his change request).
